beluga/include/beluga/views/take_evenly.hpp

```diff
@@ -30,64 +30,145 @@ namespace beluga::views {
+    [[nodiscard]] constexpr std::ptrdiff_t distance_to(const cursor& other) const {
+      return static_cast<std::ptrdiff_t>(other.pos_) - static_cast<std::ptrdiff_t>(pos_);
+    }
+
+    template <
+        class T = Range,
+        std::enable_if_t<std::is_same_v<T, Range> && ranges::random_access_range<Range>, int> = 0>
+    [[nodiscard]] constexpr std::ptrdiff_t distance_to(const ranges::default_sentinel_t&) const {
+      return static_cast<std::ptrdiff_t>(view_->count_) - static_cast<std::ptrdiff_t>(pos_);
+    }
+
+   private:
+    const take_evenly_view* view_;
+    std::size_t pos_;
+
+    ranges::iterator_t<Range> it_;
+  };
+
+  [[nodiscard]] constexpr auto begin_cursor() const { return cursor{this, 0U}; }
+
+  [[nodiscard]] constexpr auto end_cursor() const noexcept { return ranges::default_sentinel_t{}; }
+
+  [[nodiscard]] constexpr std::ptrdiff_t compute_offset(std::size_t pos) const {
+    if (pos == 0U) {
+      return 0;
+    }
+
+    if (count_ == 1U) {
+      return static_cast<std::ptrdiff_t>(size_);
+    }
+
+    const std::ptrdiff_t a = static_cast<std::ptrdiff_t>(pos) * (static_cast<std::ptrdiff_t>(size_) - 1);
+    const std::ptrdiff_t b = static_cast<std::ptrdiff_t>(count_) - 1;
+    return a / b + ((a % b == 0) ? 0 : 1);
+  }
+
+  [[nodiscard]] constexpr std::ptrdiff_t compute_offset(std::size_t current, std::size_t target) const {
+    if (count_ > size_) {
+      return static_cast<std::ptrdiff_t>(target) - static_cast<std::ptrdiff_t>(current);
+    }
+
+    return compute_offset(target) - compute_offset(current);
+  }
+
+  Range range_;
+  std::size_t count_;
+  std::size_t size_;
+};
+
+template <class Range>
+take_evenly_view(Range&& range, std::size_t max) -> take_evenly_view<ranges::views::all_t<Range>>;
+
+struct take_evenly_fn {
   template <class Range>
   constexpr auto operator()(Range&& range, std::size_t count) const {
-    const std::size_t size = ranges::size(range);
-
-    const auto filter_function = [size, count](const auto& pair) {
-      if ((size == 0UL) || (count == 0UL)) {
-        return false;
-      }
-
-      if (count > size) {
-        return true;
-      }
-
-      const auto [index, _] = pair;
-      if (count == 1UL) {
-        return index == 0UL;
-      }
-
-      if ((index == 0UL) || (index == size - 1UL)) {
-        return true;
-      }
-
-      const std::size_t m0 = (index - 1UL) * (count - 1UL) / (size - 1UL);
-      const std::size_t m1 = index * (count - 1UL) / (size - 1UL);
-      return m0 != m1;
-    };
-
-    // `cache1` ensures that views prior to `filter` in the pipeline are iterated exactly once.
-    // This is needed because filter needs to dereference the input iterator twice.
-    return ranges::views::enumerate(range) | ranges::views::cache1 | ranges::views::filter(filter_function) |
-           beluga::views::elements<1>;
+    return take_evenly_view{std::forward<Range>(range), count};
```

Comment:
Slightly helps with #385 by reducing the max RSS when compiling using a single thread.

Before: Maximum resident set size (kbytes): 2703864
After: Maximum resident set size (kbytes): 2640764